File: Blind75-Practice/easy/14-math-geometry/Python/greatest-common-divisor-of-strings.py
"""
LeetCode 1071: Greatest Common Divisor of Strings
Difficulty: Easy
Pattern: Math & Geometry

Largest string x such that str1 and str2 can be formed by repeating x.

Key fact: a common divisor exists iff str1 + str2 == str2 + str1. Then answer length is
gcd(len(str1), len(str2)).

Time Complexity: O(len(str1) + len(str2)) for concatenation check + gcd
Space Complexity: O(len(str1) + len(str2)) for the concatenations (can avoid full concat for
check by comparing in passes; kept simple here)
"""

import math


class Solution:
    def gcdOfStrings(self, str1: str, str2: str) -> str:
        if str1 + str2 != str2 + str1:
            return ""
        g = math.gcd(len(str1), len(str2))
        return str1[:g]


if __name__ == "__main__":
    s = Solution()
    assert s.gcdOfStrings("ABCABC", "ABC") == "ABC"
    assert s.gcdOfStrings("ABABAB", "ABAB") == "AB"
    assert s.gcdOfStrings("LEET", "CODE") == ""
    print("All test cases passed!")

File: Blind75-Practice/easy/18-math-geometry-l/Java/GreatestCommonDivisorOfStrings.java
/*
 * 1071. Greatest Common Divisor of Strings
 * Difficulty: Easy
 * Pattern: Math & Geometry
 *
 * Problem:
 * For two strings `str1` and `str2`, we say `t` divides a string `s` if `s` is formed
 * by concatenating `t` one or more times.
 *
 * Return the largest string `x` such that `x` divides both `str1` and `str2`.
 *
 * Example 1:
 * Input: str1 = "ABCABC", str2 = "ABC"
 * Output: "ABC"
 *
 * Example 2:
 * Input: str1 = "ABABAB", str2 = "ABAB"
 * Output: "AB"
 *
 * Example 3:
 * Input: str1 = "LEET", str2 = "CODE"
 * Output: ""
 *
 * Constraints:
 * - 1 <= str1.length, str2.length <= 1000
 * - str1 and str2 consist of uppercase English letters
 *
 * Notes:
 * - If `str1 + str2` is not equal to `str2 + str1`, then no common divisor string exists.
 * - Otherwise the answer length is `gcd(str1.length(), str2.length())`.
 */

public class GreatestCommonDivisorOfStrings {

    private static int gcd(int a, int b) {
    }

    public String gcdOfStrings(String str1, String str2) {
        //

        //

        //

        //

        //

        //

        //

        //

        //

        //

        //

        //

        //

        //

        //

        //

        //

        //

        //

        //

        //

        //

        //

        //

        //

        //

        //

        //

        //

        //

        //

        //

        //

        //

        //

        //

        //

        //

        //

        //

        // private static int gcd(int a, int b) {
        //     while (b != 0) {
        //         int temp = a % b;
        //         a = b;
        //         b = temp;
        //     }
        //     return a;
        // }
        //
        // public String gcdOfStrings(String str1, String str2) {
        //     if (!(str1 + str2).equals(str2 + str1)) {
        //         return "";
        //     }
        //     int length = gcd(str1.length(), str2.length());
        //     return str1.substring(0, length);
        // }
    }

    public static void main(String[] args) {
        GreatestCommonDivisorOfStrings s = new GreatestCommonDivisorOfStrings();
        assert s.gcdOfStrings("ABCABC", "ABC").equals("ABC");
        assert s.gcdOfStrings("ABABAB", "ABAB").equals("AB");
        assert s.gcdOfStrings("LEET", "CODE").equals("");
        System.out.println("All test cases passed!");
    }
}
